File: Go/GORM_ORM/dal/repositories/shipperRepository.go
package repositories

import (
	"GORM_ORM/dal/entities"

	"gorm.io/gorm"
)

type ShipperRepository interface {
	GetAll() ([]entities.Shipper, error)
	GetByID(id int) (*entities.Shipper, error)
	Create(entity *entities.Shipper) error
	Update(entity *entities.Shipper) error
	Delete(id int) error
}

type shipperRepository struct {
	genericRepo GenericRepository[entities.Shipper]
}

func NewShipperRepository(db *gorm.DB) ShipperRepository {
	return &shipperRepository{
		genericRepo: NewGenericRepository[entities.Shipper](db),
	}
}

func (r *shipperRepository) GetAll() ([]entities.Shipper, error) {
	return r.genericRepo.GetAll()
}

func (r *shipperRepository) GetByID(id int) (*entities.Shipper, error) {
	return r.genericRepo.GetByID(id)
}

func (r *shipperRepository) Create(entity *entities.Shipper) error {
	return r.genericRepo.Create(entity)
}

func (r *shipperRepository) Update(entity *entities.Shipper) error {
	return r.genericRepo.Update(entity)
}

func (r *shipperRepository) Delete(id int) error {
	return r.genericRepo.Delete(id)
}

File: C#/Dapper_ORM/DAL/Repositories/ShipperRepository.cs
using DAL.Entities;
using Dapper;
using System.Data.SqlClient;

namespace DAL.Repositories;

public class ShipperRepository /*: GenericRepository<Shipper, int>*/
{
    private readonly string _connectionString;
    public ShipperRepository(string connectionString) /*: base(connectionString, "Shippers", "ShipperID")*/
    {
        _connectionString = connectionString;
    }
    [Obsolete]
    public async Task<List<Shipper>> GetAllAsync()
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            await connection.OpenAsync();
            var query = $"SELECT * FROM Shippers";
            var result = await connection.QueryAsync<Shipper>(query);
            return result.ToList();
        }
    }
    [Obsolete]
    public async Task<Shipper?> GetByIdAsync(int id)
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            connection.Open();
            var query = $"SELECT * FROM Shippers WHERE ShipperID = @Id";
            return await connection.QueryFirstOrDefaultAsync<Shipper>(query, new { Id = id });
        }
    }
    [Obsolete]
    public async Task<Shipper> AddAsync(Shipper entity)
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            await connection.OpenAsync();
            var query = $"INSERT INTO Shippers (CompanyName, Phone) VALUES (@CompanyName, @Phone); SELECT CAST(SCOPE_IDENTITY() as int);";
            var id = await connection.QuerySingleAsync<int>(query, entity);
            var propertyInfo = typeof(Shipper).GetProperty("ShipperID");
            propertyInfo.SetValue(entity, id);
            return entity;
        }
    }
    [Obsolete]
    public async Task UpdateAsync(Shipper entity)
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            await connection.OpenAsync();
            var query = $"UPDATE Shippers SET CompanyName = @CompanyName, Phone = @Phone WHERE ShipperID = @ShipperID";
            await connection.ExecuteAsync(query, entity);
        }
    }
    [Obsolete]
    public async Task DeleteAsync(int id)
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            await connection.OpenAsync();
            var query = $"DELETE FROM Shippers WHERE ShipperID = @Id";
            await connection.ExecuteAsync(query, new { Id = id });
        }
    }
}
